community/planning/ids_as_rust_structs.md

```diff
@@ -0,0 +1,306 @@
+which parse the strings, a side-effect is the partial validation of the string
+which potentially results in an error (though currently the return of
+`InvalidArgumentError` is not consistent, because this error was introduced
+after much of the code was written). The parsing is often ad-hoc. By using
+structs, the opportunity for these runtime errors is removed from all code
+except for code which calls the struct's constructors. By creating the structs
+early and using them throughout the rest of the code, the opportunity for
+runtime parsing errors in low-level functions is removed. Parsing the string
+multiple times is also avoided.
+
+## Guide-level Explanation
+
+### Circuit Identifiers
+
+A valid circuit identifier is a string of length 11 with the following
+structure:
+
+* characters 0-4 are alphanumeric
+* character 5 is a `-`
+* characters 6-10 are alphanumeric
+
+A circuit identifier string may be converted to an integer by removing the `-`
+character and then using base62 conversion. Likewise, the reverse operation can
+be performed to convert an integer to a circuit identifier string.
+
+To represent a circuit identifier, we only need a single struct called
+`CircuitId`.
+
+### Service Identifiers
+
+There are two types of service identifiers, each with different formats.
+
+A typical service id consists of a string of length 4, with all characters
+being alphanumeric. These service ids can be converted to an integer and back
+using base62 encoding.
+
+An admin service id, which is used for routing to admin services, starts with
+the string `admin::` and can have one of two forms:
+
+* `admin::<node_id>`
+* `admin::public_key::<remote_public_key>::public_key::<local_public_key>`
+
+
+A `<node_id>` consists of any valid utf-8 string. This will change to a more
+restrictive string in the future to prevent ambiguous parsing of service
+identifiers and other potential issues with undesirable characters like spaces.
+
+The public key substrings `<remote_public_key>` and `<local_public_key>`
+consist of hexadecimal characters.
+
+To represent any valid service identifier, we use a struct called `ServiceId`.
+Additionally, to support APIs which can only work with admin service
+identifiers, we provide a struct called `AdminServiceId` which is restricted to
+only service identifiers which start with `admin::`.
+
+The admin service id cannot be represented as an integer. Thus as a whole,
+service id cannot be represented as an integer.
+
+### Fully Qualified Service Identifiers
+
+It is common to combine circuit and service identifiers into a single string,
+of the format `<circuit_id>::<service_id>`. This is called a fully-qualified
+service identifier and is supported with the struct `FullyQualifiedServiceId`.
+
+### Examples
+
+The following are example valid circuit ids:
+
+* `ABCDE-01234`
+* `foA8k-03kAM`
+
+The following are examples of valid service ids:
+
+* `00aa`
+* `45R3`
+* `Amrk`
+* `admin::node1`

```

Comment:
this would just be `node1` as the `admin` is technically the circuit, the admin circuit